### PR TITLE
[SINT-1180] Index Allowlist

### PR DESCRIPTION
**Pull request by cedricvanrompay-datadog**
See https://datadoghq.atlassian.net/browse/SINT-1180

### CODE REVIEW

File: yorkshire/cli.py [hunk 1]
```diff
@@ -49,9 +50,14 @@ def cli(debug: bool = False) -> None:
     type=str,
     metavar="FILE|URL|DIR",
 )
-def cli_detect(ctx: click.core.Context, path: str) -> None:
+@click.option(
+    '--allowed-index',
```
Comment (fridex) [inline yorkshire/cli.py:54]:
```suggestion
    '--allowed-index-url',
```

File: yorkshire/cli.py [hunk 2]
```diff
@@ -49,9 +50,14 @@ def cli(debug: bool = False) -> None:
...
+    multiple=True,
+    help="allow this particular index",
+)
+def cli_detect(ctx: click.core.Context, path: str, allowed_indexes: Iterable[str]) -> None:
```
Comment (fridex) [inline yorkshire/cli.py:58]: Name of the argument needs to match the CLI option:

```suggestion
def cli_detect(ctx: click.core.Context, path: str, allowed_index_url: Tuple[str]) -> None:
```

Also, click passes a `Tuple` in this case.

File: yorkshire/cli.py [hunk 3]
```diff
@@ -49,9 +50,14 @@ def cli(debug: bool = False) -> None:
...
     """Check for a possible dependency confusion in a requirements file, files in a directory, or a URL."""
-    okay = all([i[1] for i in yorkshire.detect(path)])
+    okay = all([i[1] for i in yorkshire.detect(path, allowed_indexes)])
```
Comment (fridex) [inline yorkshire/cli.py:60]:
```suggestion
    okay = all([i[1] for i in yorkshire.detect(path, allowed_index_url)])
```

File: yorkshire/cli.py [hunk 4]
```diff
@@ -1,6 +1,7 @@
 #!/usr/bin/env python3
 
 import logging
+from typing import Iterable
```
Comment (fridex) [inline yorkshire/cli.py:4]:
```suggestion
from typing import Tuple
```

File: tests/test_detect_file.py [hunk 5]
```diff
@@ -26,7 +30,16 @@ def test_okay(self, filepath: str) -> None:
             f"Test case for {filepath!r} should NOT be vulnerable to dependency confusion"
 
     @pytest.mark.parametrize("filepath", list(_iter_files(os.path.join(Base.DATA_DIR, "requirements_files", "fail"))))
-    def test_fail(self, filepath: str) -> None:
+    def test_fail(self, filepath: str, index_allowlist: bool) -> None:
```
Comment (fridex) [inline tests/test_detect_file.py:33]: Could `index_allowlist` be part of the `pytest.mark.parametrize`? Other tests use it, and it looks kind of weird to combine both. Also, `ALLOWED_INDEXES` could become part of the `pytest.mark.parametrize` as well.

File: yorkshire/_lib.py [hunk 6]
```diff
@@ -107,8 +132,14 @@ def _detect_pipfile(filepath: str, *, _real_path: Optional[str] = None) -> bool:
     except Exception as exc:
         raise FileParseError(f"Failed to open and parse Pipfile file {real_filepath!r}: {str(exc)}") from exc
 
-    sources = content.get("source") or []
-    if len(sources) > 1:
+    sources = [
+        source for source in content.get("source") or []
+        if not allowed_indexes or source["url"] not in allowed_indexes
+    ]
```
Comment (fridex) [inline yorkshire/_lib.py:138]: Could this be rewritten to a more readable form (also below)?

**Review by fridex — CHANGES_REQUESTED** [review]

**Comment by cedricvanrompay-datadog** [comment]
@fridex should be all done

**Review by fridex — APPROVED** [review]

**Pull request merged**